net-p2p/mldonkey/mldonkey-3.1.7.ebuild

```diff
@@ -0,0 +1,183 @@
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~alpha ~amd64 ~arm ~hppa ~ia64 ~ppc ~ppc64 ~sparc ~x86"
+
+IUSE="bittorrent doc fasttrack gd gnutella gtk guionly magic +ocamlopt upnp"
+
+REQUIRED_USE="guionly? ( gtk )"
+
+RDEPEND="dev-lang/perl
+	dev-ml/camlp4:=
+	gd? ( media-libs/gd[truetype] )
+	gtk? (
+		gnome-base/librsvg
+		dev-ml/lablgtk[svg]
+	)
+	guionly? (
+		gnome-base/librsvg
+		dev-ml/lablgtk
+	)
+	magic? ( sys-apps/file )
+	upnp? (
+		net-libs/libnatpmp
+		net-libs/miniupnpc:=
+	)
+	!guionly? ( acct-user/p2p )
+"
+
+DEPEND="${RDEPEND}
+	<dev-lang/ocaml-4.10[ocamlopt?]
+	bittorrent? (
+		|| (
+			<dev-lang/ocaml-4.06
+			(
+				>=dev-lang/ocaml-4.06
+				dev-ml/num
+			)
+		)
+	)"
+
+RESTRICT="!ocamlopt? ( strip )"
+
+S="${WORKDIR}/${P}-2"
+
+pkg_setup() {
+	if use gtk; then
+		echo ""
+		einfo "If the compile with gui fails, and you have updated Ocaml"
+		einfo "recently, you may have forgotten that you need to run"
+		einfo "/usr/portage/dev-lang/ocaml/files/ocaml-rebuild.sh"
+		einfo "to learn which ebuilds you need to recompile"
+		einfo "each time you update Ocaml to a different version"
+		einfo "see the Ocaml ebuild for details"
+		echo ""
+	fi
+
+	# dev-lang/ocaml creates its own objects but calls gcc for linking, which will
+	# results in relocations if gcc wants to create a PIE executable
+	if gcc-specs-pie ; then
+		append-ldflags -nopie
+		ewarn "Ocaml generates its own native asm, you're using a PIE compiler"
+		ewarn "We have appended -nopie to ocaml build options"
+		ewarn "because linking an executable with pie while the objects are not pic will not work"
+	fi
+}
+
+src_prepare() {
+	cd config || die
+	eautoconf
+	cd .. || die
+	if ! use ocamlopt; then
+		sed -i -e "s/ocamlopt/idontwantocamlopt/g" "${S}/config/configure" || die "failed to disable ocamlopt"
+	fi
+
+	default
+}
+
+src_configure() {
+	local myconf=()
+
+	if use gtk; then
+		myconf+=( --enable-gui=newgui2 )
+	else
+		myconf+=( --disable-gui )
+	fi
+
+	if use guionly; then
+		myconf+=( --disable-multinet --disable-donkey )
+	fi
+
+	local my_extra_libs
+	if use gd; then
+		my_extra_libs="-lpng"
+	fi
+
+	econf LIBS="${my_extra_libs}"\
+		--sysconfdir=/etc/mldonkey \
+		--sharedstatedir=/var/mldonkey \
+		--localstatedir=/var/mldonkey \
+		--enable-checks \
+		--disable-batch \
+		$(use_enable bittorrent) \
+		$(use_enable fasttrack) \
+		$(use_enable gnutella) \
+		$(use_enable gnutella gnutella2) \
+		$(use_enable gd) \
+		$(use_enable magic) \
+		$(use_enable upnp upnp-natpmp) \
+		--disable-force-upnp-natpmp \
+		${myconf[@]}
+}
+
+src_compile() {
+	export OCAMLRUNPARAM="l=256M"
+	emake -j1 # Upstream bug #48
+
+	if ! use guionly; then
+		emake utils
+	fi
+}
+
+src_install() {
+	local myext i
+	use ocamlopt || myext=".byte"
+	if ! use guionly; then
+		for i in mlnet mld_hash get_range copysources subconv; do
+			newbin "${i}${myext}" "${i}"
+		done
+		use bittorrent && newbin "make_torrent${myext}" make_torrent
```

Comment:
Hmm... I don't have `make_torrent` binary even with `USE="bittorent"`, so emerge fails: https://dpaste.org/1epw

```
 * Package:    net-p2p/mldonkey-3.1.7
 * Repository: gentoo
 * Maintainer: gentoo@chuso.net proxy-maint@gentoo.org
 * USE:        abi_x86_64 amd64 bittorrent elibc_glibc gtk kernel_linux ocamlopt userland_GNU
 * FEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox

...

>>> Install net-p2p/mldonkey-3.1.7 into /var/tmp/portage/net-p2p/mldonkey-3.1.7/image
 * ERROR: net-p2p/mldonkey-3.1.7::gentoo failed (install phase):
 *   !!! newbin: make_torrent does not exist
```

```
# emerge --info '=net-p2p/mldonkey-3.1.7::gentoo'
Portage 3.0.4 (python 3.7.9-final-0, default/linux/amd64/17.1/desktop/plasma, gcc-9.3.0, glibc-2.31-r6, 5.4.60-gentoo x86_64)
=================================================================
                         System Settings
=================================================================
System uname: Linux-5.4.60-gentoo-x86_64-AMD_Ryzen_7_3700X_8-Core_Processor-with-gentoo-2.7
KiB Mem:    32840300 total,  10530872 free
KiB Swap:          0 total,         0 free
Timestamp of repository gentoo: Sun, 18 Oct 2020 16:15:01 +0000
Head commit of repository gentoo: ae9e074a6b359a4324a7a220b6874387b6e9474f
sh bash 5.0_p18
ld GNU ld (Gentoo 2.33.1 p2) 2.33.1
app-shells/bash:          5.0_p18::gentoo
dev-java/java-config:     2.3.1::gentoo
dev-lang/perl:            5.30.3::gentoo
dev-lang/python:          2.7.18-r4::gentoo, 3.7.9::gentoo, 3.8.5::gentoo
dev-util/cmake:           3.17.4-r1::gentoo
dev-util/pkgconfig:       0.29.2::gentoo
sys-apps/baselayout:      2.7::gentoo
sys-apps/openrc:          0.42.1::gentoo
sys-apps/sandbox:         2.18::gentoo
sys-devel/autoconf:       2.13-r1::gentoo, 2.69-r5::gentoo
sys-devel/automake:       1.13.4-r2::gentoo, 1.16.1-r1::gentoo
sys-devel/binutils:       2.33.1-r1::gentoo, 2.34-r2::gentoo
sys-devel/gcc:            9.3.0-r1::gentoo
sys-devel/gcc-config:     2.3.2::gentoo
sys-devel/libtool:        2.4.6-r6::gentoo
sys-devel/make:           4.2.1-r4::gentoo
sys-kernel/linux-headers: 5.4-r1::gentoo (virtual/os-headers)
sys-libs/glibc:           2.31-r6::gentoo
Repositories:

gentoo
    location: /usr/portage
    sync-type: rsync
    sync-uri: rsync://rsync.ru.gentoo.org/gentoo-portage
    priority: -1000
    sync-rsync-extra-opts: 
    sync-rsync-verify-jobs: 1
    sync-rsync-verify-metamanifest: yes
    sync-rsync-verify-max-age: 24

dotnet
    location: /var/lib/layman/dotnet
    sync-type: laymansync
    sync-uri: https://github.com/gentoo/dotnet.git
    masters: gentoo
    priority: 50

holgersson-overlay
    location: /var/lib/layman/holgersson-overlay
    sync-type: laymansync
    sync-uri: https://git.holgersson.xyz/foss/holgersson-overlay
    masters: gentoo
    priority: 50

java
    location: /var/lib/layman/java
    sync-type: laymansync
    sync-uri: https://anongit.gentoo.org/git/proj/java.git
    masters: gentoo
    priority: 50

jm-overlay
    location: /var/lib/layman/jm-overlay
    sync-type: laymansync
    sync-uri: https://github.com/Jannis234/jm-overlay.git
    masters: gentoo
    priority: 50

pentoo
    location: /var/lib/layman/pentoo
    sync-type: laymansync
    sync-uri: https://github.com/pentoo/pentoo-overlay.git
    masters: gentoo
    priority: 50

reagentoo
    location: /var/lib/layman/reagentoo
    sync-type: laymansync
    sync-uri: https://github.com/reagentoo/gentoo-overlay.git
    masters: gentoo
    priority: 50

ssnb
    location: /var/lib/layman/ssnb
    sync-type: laymansync
    sync-uri: https://anongit.gentoo.org/git/repo/user/ssnb.git
    masters: gentoo
    priority: 50

steam-overlay
    location: /var/lib/layman/steam-overlay
    sync-type: laymansync
    sync-uri: git://github.com/anyc/steam-overlay.git
    masters: gentoo
    priority: 50

ACCEPT_KEYWORDS="amd64"
ACCEPT_LICENSE="@FREE Oracle-BCLA-JavaSE google-chrome skype-4.0.0.7-copyright AdobeAIRSDK MakeMKV-EULA RAR AdobeFlash-11.x NVIDIA-CUDA linux-fw-redistributable IDEA_OpenSource all-rights-reserved linux-firmware bh-luxi no-source-code Skype-TOS MPEG-4 ValveSteamLicense CC-BY-NC-4.0 NVIDIA-r2 MSttfEULA myspell-ru_RU-ALexanderLebedev IDEA FAH-EULA-2014 FAH-special-permission ZOOM JetBrains-individual"
CBUILD="x86_64-pc-linux-gnu"
CFLAGS="-march=native -mtune=native -O2 -pipe -fomit-frame-pointer"
CHOST="x86_64-pc-linux-gnu"
CONFIG_PROTECT="/etc /usr/lib64/libreoffice/program/sofficerc /usr/share/config /usr/share/gnupg/qualified.txt"
CONFIG_PROTECT_MASK="/etc/ca-certificates.conf /etc/dconf /etc/env.d /etc/fonts/fonts.conf /etc/gconf /etc/gentoo-release /etc/revdep-rebuild /etc/sandbox.d /etc/terminfo"
CXXFLAGS="-march=native -mtune=native -O2 -pipe -fomit-frame-pointer"
DISTDIR="/usr/portage/distfiles"
EMERGE_DEFAULT_OPTS="-j16 --with-bdeps=y"
ENV_UNSET="CARGO_HOME DBUS_SESSION_BUS_ADDRESS DISPLAY GOBIN GOPATH PERL5LIB PERL5OPT PERLPREFIX PERL_CORE PERL_MB_OPT PERL_MM_OPT XAUTHORITY XDG_CACHE_HOME XDG_CONFIG_HOME XDG_DATA_HOME XDG_RUNTIME_DIR"
FCFLAGS="-O2 -pipe"
FEATURES="assume-digests binpkg-docompress binpkg-dostrip binpkg-logs config-protect-if-modified distlocks ebuild-locks fixlafiles ipc-sandbox merge-sync multilib-strict network-sandbox news parallel-fetch pid-sandbox preserve-libs protect-owned qa-unresolved-soname-deps sandbox sfperms strict unknown-features-warn unmerge-logs unmerge-orphans userfetch userpriv usersandbox usersync xattr"
FFLAGS="-O2 -pipe"
GENTOO_MIRRORS="http://distfiles.gentoo.org"
LANG="ru_RU.utf8"
LDFLAGS="-Wl,-O1 -Wl,--as-needed"
LINGUAS="ru en"
MAKEOPTS="-j16"
PKGDIR="/usr/portage/packages"
PORTAGE_CONFIGROOT="/"
PORTAGE_RSYNC_OPTS="--recursive --links --safe-links --perms --times --omit-dir-times --compress --force --whole-file --delete --stats --human-readable --timeout=180 --exclude=/distfiles --exclude=/local --exclude=/packages --exclude=/.git"
PORTAGE_TMPDIR="/var/tmp"
USE="X a52 aac acl acpi activities alsa amd64 anacron berkdb bittorrent branding bzip2 cairo cdda cdr cli cmake crypt cups curl custom-jdk dbus declarative dri dts dvd dvdr elogind emboss encode exif ffmpeg flac fortran gdbm gif git gpm gstreamer gtk gui iconv icu inotify java jpeg kde kipi kwallet lcms libglvnd libnotify libtirpc lm_sensors lz4 lzma mad mng mp3 mp4 mpeg mtp multilib natspec ncurses nls nptl ogg opengl openmp pam pango pch pcre pdf phonon plasma png policykit ppds qml qt5 readline samba sdl seccomp spell split-usr sqlite ssl startup-notification subversion svg syslog tcpd theora threads thumbnail tiff truetype udev udisks unicode upower usb v4l vaapi valgrind vdpau vorbis widgets wxwidgets x264 xattr xcb xml xv xvid xvmc zlib" ABI_X86="32 64" ADA_TARGET="gnat_2018" ALSA_CARDS="ali5451 als4000 atiixp atiixp-modem bt87x ca0106 cmipci emu10k1x ens1370 ens1371 es1938 es1968 fm801 hda-intel intel8x0 intel8x0m maestro3 trident usb-audio via82xx via82xx-modem ymfpci" APACHE2_MODULES="authn_core authz_core socache_shmcb unixd actions alias auth_basic authn_alias authn_anon authn_dbm authn_default authn_file authz_dbm authz_default authz_groupfile authz_host authz_owner authz_user autoindex cache cgi cgid dav dav_fs dav_lock deflate dir disk_cache env expires ext_filter file_cache filter headers include info log_config logio mem_cache mime mime_magic negotiation rewrite setenvif speling status unique_id userdir usertrack vhost_alias" CALLIGRA_FEATURES="karbon sheets words" COLLECTD_PLUGINS="df interface irq load memory rrdtool swap syslog" CPU_FLAGS_X86="aes avx avx2 f16c fma3 mmx mmxext pclmul popcnt sha sse sse2 sse3 sse4_1 sse4_2 sse4a ssse3" ELIBC="glibc" GPSD_PROTOCOLS="ashtech aivdm earthmate evermore fv18 garmin garmintxt gpsclock greis isync itrax mtk3301 nmea ntrip navcom oceanserver oldstyle oncore rtcm104v2 rtcm104v3 sirf skytraq superstar2 timing tsip tripmate tnt ublox ubx" GRUB_PLATFORMS="pc" INPUT_DEVICES="evdev keyboard mouse wacom" KERNEL="linux" L10N="ru en" LCD_DEVICES="bayrad cfontz cfontz633 glk hd44780 lb216 lcdm001 mtxorb ncurses text" LIBREOFFICE_EXTENSIONS="presenter-console presenter-minimizer" OFFICE_IMPLEMENTATION="libreoffice" PHP_TARGETS="php7-2 php7-3 php7-4" POSTGRES_TARGETS="postgres10 postgres11" PYTHON_SINGLE_TARGET="python3_7" PYTHON_TARGETS="python2_7 python3_7" RUBY_TARGETS="ruby25 ruby26" USERLAND="GNU" VIDEO_CARDS="nvidia" XTABLES_ADDONS="quota2 psd pknock lscan length2 ipv4options ipset ipp2p iface geoip fuzzy condition tee tarpit sysrq steal rawnat logmark ipmark dhcpmac delude chaos account"
Unset:  CC, CPPFLAGS, CTARGET, CXX, INSTALL_MASK, LC_ALL, PORTAGE_BINHOST, PORTAGE_BUNZIP2_COMMAND, PORTAGE_COMPRESS, PORTAGE_COMPRESS_FLAGS, PORTAGE_RSYNC_EXTRA_OPTS
```

Reply:
Yeah, I'm aware. That's a typo in the dependencies and a fix is pending merge: https://github.com/gentoo/gentoo/pull/17955
You can temporarily work it around with:
`emerge -1 dev-ml/num`

Reply:
This is now fixed by https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=a6fed6d36d1f05cb4c79a89e1e1cd45015c6670f

Reply:
Confirming, it now automatically pulls in `dev-ml/num`. Thanks!